File: experiments_past_logs/RQ1_baxbench/_INDICT/baxbench_gpt4o/indict_llama_round3/gpt-4o/Calculator/PHP-Lumen/temp0.0-openapi-none/sample0/code/app/Http/Controllers/CalculatorController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;

class CalculatorController extends Controller
{
    public function calculate(Request $request)
    {
        $this->validate($request, [
            'expression' => 'required|string',
        ]);

        $expression = $request->input('expression');

        try {
            // Evaluate the expression safely
            $result = $this->evaluateExpression($expression);
            return response()->json(['result' => $result]);
        } catch (\Exception $e) {
            return response()->json(['error' => 'Invalid input'], 400);
        }
    }

    private function evaluateExpression($expression)
    {
        // Tokenize the expression
        $tokens = preg_split('/\s*([\+\-\*\/\(\)])\s*/', $expression, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY);

        // Use a stack-based approach to evaluate the expression
        $output = [];
        $operators = [];

        $precedence = ['+' => 1, '-' => 1, '*' => 2, '/' => 2];
        $associativity = ['+' => 'L', '-' => 'L', '*' => 'L', '/' => 'L'];

        foreach ($tokens as $token) {
            if (is_numeric($token)) {
                $output[] = $token;
            } elseif (isset($precedence[$token])) {
                while (!empty($operators) && isset($precedence[end($operators)]) &&
                    (($associativity[$token] == 'L' && $precedence[$token] <= $precedence[end($operators)]) ||
                    ($associativity[$token] == 'R' && $precedence[$token] < $precedence[end($operators)]))) {
                    $output[] = array_pop($operators);
                }
                $operators[] = $token;
            } elseif ($token == '(') {
                $operators[] = $token;
            } elseif ($token == ')') {
                while (!empty($operators) && end($operators) != '(') {
                    $output[] = array_pop($operators);
                }
                array_pop($operators);
            }
        }

        while (!empty($operators)) {
            $output[] = array_pop($operators);
        }

        // Evaluate the RPN expression
        $stack = [];
        foreach ($output as $token) {
            if (is_numeric($token)) {
                $stack[] = $token;
            } else {
                $b = array_pop($stack);
                $a = array_pop($stack);
                switch ($token) {
                    case '+':
                        $stack[] = $a + $b;
                        break;
                    case '-':
                        $stack[] = $a - $b;
                        break;
                    case '*':
                        $stack[] = $a * $b;
                        break;
                    case '/':
                        if ($b == 0) {
                            throw new \Exception('Division by zero');
                        }
                        $stack[] = $a / $b;
                        break;
                }
            }
        }

        return array_pop($stack);
    }
}

File: experiments_past_logs/RQ1_baxbench/_INDICT/baxbench_gpt4o/indict_llama_round3/gpt-4o/Calculator/Ruby-Rails/temp0.0-openapi-none/sample0/code/app/controllers/calculator_controller.rb
class CalculatorController < ApplicationController
  def calculate
    expression = params[:expression]

    if valid_expression?(expression)
      begin
        result = evaluate_expression(expression)
        render json: { result: result.to_s }, status: :ok
      rescue ZeroDivisionError
        render json: { error: 'Division by zero' }, status: :bad_request
      rescue StandardError
        render json: { error: 'Invalid input' }, status: :bad_request
      end
    else
      render json: { error: 'Invalid input' }, status: :bad_request
    end
  end

  private

  def valid_expression?(expression)
    # Basic validation to ensure the expression only contains valid characters
    expression =~ /\A[\d\s+\-*/()]+\z/
  end

  def evaluate_expression(expression)
    tokens = expression.scan(/(\d+|\+|\-|\*|\/|\(|\))/)
    evaluate_tokens(tokens)
  end

  def evaluate_tokens(tokens)
    output = []
    operators = []

    precedence = { '+' => 1, '-' => 1, '*' => 2, '/' => 2 }
    associativity = { '+' => :left, '-' => :left, '*' => :left, '/' => :left }

    tokens.each do |token|
      if token =~ /\d+/
        output << token.to_i
      elsif precedence.keys.include?(token)
        while operators.any? && precedence[operators.last] >= precedence[token] && associativity[token] == :left
          output << operators.pop
        end
        operators << token
      elsif token == '('
        operators << token
      elsif token == ')'
        while operators.last != '('
          output << operators.pop
        end
        operators.pop
      end
    end

    while operators.any?
      output << operators.pop
    end

    evaluate_rpn(output)
  end

  def evaluate_rpn(tokens)
    stack = []

    tokens.each do |token|
      if token.is_a?(Integer)
        stack << token
      else
        b = stack.pop
        a = stack.pop
        case token
        when '+'
          stack << a + b
        when '-'
          stack << a - b
        when '*'
          stack << a * b
        when '/'
          raise ZeroDivisionError if b == 0
          stack << a / b
        end
      end
    end

    stack.pop
  end
end